crates/wal-protocol/src/lib.rs

```diff
@@ -225,3 +273,274 @@ impl MatchKeyQuery for Envelope {
+                CommandKind::InvokerEffect,
+                Field::encode_serde(StorageCodecKind::FlexbuffersSerde, value),
+            ),
+            Command::Timer(value) => (
+                CommandKind::Timer,
+                Field::encode_serde(StorageCodecKind::FlexbuffersSerde, value),
+            ),
+            Command::ScheduleTimer(value) => (
+                CommandKind::ScheduleTimer,
+                Field::encode_serde(StorageCodecKind::FlexbuffersSerde, value),
+            ),
+            Command::InvocationResponse(value) => {
+                let value =
+                    protobuf::outbox_message::OutboxServiceInvocationResponse::from(value.clone());
+                (
+                    CommandKind::InvocationResponse,
+                    Field::encode_protobuf(&value),
+                )
+            }
+            Command::NotifyGetInvocationOutputResponse(value) => (
+                CommandKind::NotifyGetInvocationOutputResponse,
+                Field::encode_bilrost(value),
+            ),
+            Command::NotifySignal(value) => {
+                let value = protobuf::outbox_message::NotifySignal::from(value.clone());
+                (CommandKind::NotifySignal, Field::encode_protobuf(&value))
+            }
+        };
+
+        let dto = Envelope {
+            header: Field::encode_serde(StorageCodecKind::FlexbuffersSerde, &envelope.header)?,
+            command_kind,
+            command: command?,
+        };
+
+        Ok(dto.encode_fast().into_vec().into())
+    }
+
+    pub fn decode<B: Buf>(buf: B) -> Result<super::Envelope, StorageDecodeError> {
+        let envelope =
+            Envelope::decode(buf).map_err(|err| StorageDecodeError::DecodeValue(err.into()))?;
+
+        // header is encoded with serde
+        let header = envelope.header.decode_serde::<super::Header>()?;
+
+        let command = match envelope.command_kind {
+            CommandKind::Unknown => return Err(DecodeError::UnknownCommandKind.into()),
+            CommandKind::AnnounceLeader => {
+                Command::AnnounceLeader(envelope.command.decode_serde()?)
+            }
+            CommandKind::PatchState => {
+                let value: protobuf::StateMutation = envelope.command.decode_protobuf()?;
+                Command::PatchState(value.try_into()?)
+            }
+            CommandKind::TerminateInvocation => {
+                Command::TerminateInvocation(envelope.command.decode_bilrost()?)
+            }
+            CommandKind::PurgeInvocation => {
+                Command::PurgeInvocation(envelope.command.decode_bilrost()?)
+            }
+            CommandKind::Invoke => {
+                let value: protobuf::ServiceInvocation = envelope.command.decode_protobuf()?;
+                Command::Invoke(value.try_into()?)
+            }
+            CommandKind::TruncateOutbox => {
+                Command::TruncateOutbox(envelope.command.decode_serde()?)
```

Comment:
How do you envision evolving the codec of the variants over time if we are not including a "codec tag" along with it. For instance, If we started serializing truncate as bilrost, or would we change the top-level codec kind on every variant change?

Reply:
We do include the `codec tag` along with it. I am only not checking it during decode at the moment because I don't need it. It will only make sense to check this tag when we are actually going to change it.

Reply:
But if we are decoding messages that was written by a future version, wouldn't we want to tell the user that we can't decode for this reason rather than blindly attempting to deserialize and hoping for the best?

Reply:
Ah okay, I see what you mean